platforms/software/dependency-management/src/main/java/org/gradle/internal/component/AbstractVariantSelectionException.java

```diff
@@ -16,17 +16,46 @@
 
 package org.gradle.internal.component;
 
+import com.google.common.collect.ImmutableList;
+import org.gradle.api.internal.DocumentationRegistry;
+import org.gradle.internal.exceptions.ResolutionProvider;
 import org.gradle.internal.exceptions.StyledException;
 
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+
 /**
  * Abstract base class for all attribute matching selection failures occurring at any stage of dependency resolution.
  */
-public abstract class AbstractVariantSelectionException extends StyledException {
-    public AbstractVariantSelectionException(String message) {
-        super(message);
+public abstract class AbstractVariantSelectionException extends StyledException implements ResolutionProvider {
+    public static final String DEFAULT_MESSAGE_PREFIX = "Review the variant matching algorithm at ";
+
+    private final List<String> resolutions = new ArrayList<>(1); // Usually there is only one resolution
+
+    public AbstractVariantSelectionException(String message, DocumentationRegistry documentationRegistry) {
+        this(message, documentationRegistry, null);
     }
 
-    public AbstractVariantSelectionException(String message, Throwable cause) {
+    public AbstractVariantSelectionException(String message, DocumentationRegistry documentationRegistry, @Nullable Throwable cause) {
```

Comment:
Something is very smelly here.  Can we avoid passing `DocumentationRegistry` everywhere? 

I think it's OK for the exceptions to be dumb containers and the message and resolutions be provided externally.

This means instead of passing `DocumentationRegistry` directly into the exception to create a resolution, we create the resolution somewhere else and pass in the String. We already have `DocumentationRegistry` in `ResolutionFailureHandler` and we're creating exceptions there.  Let's just create the resolution there too.

If we have a common resolution we want to attach everywhere, I think we can live with some small duplication.

Reply:
Done.